File: python/14501.py
# 다른 사람 코드 보고 품
# 반복적으로 봐야겠음!

n = int(input())
t = []
p = []
dp = []
for i in range(n):
    a, b = map(int, input().split())
    t.append(a)
    p.append(b)
    dp.append(b)
dp.append(0)
for i in range(n-1, -1, -1):
    if t[i] + i > n:
        dp[i] = dp[i+1]
    else:
        dp[i] = max(dp[i+1], p[i]+dp[t[i]+i])
print(dp[0])

File: c++/14501.cpp
#include <iostream>

using namespace std;

int dp[16];
int t[16];
int p[16];

int main(){
  int n;
  cin >> n;

  for(int i=1; i<=n; i++){
    cin >> t[i] >> p[i];
  }

  for(int i=n; i>=1; i--){
    if(t[i]+i-1 > n){
      dp[i] = dp[i+1];
    }else{
      dp[i] = max(dp[i+1], dp[i+t[i]]+p[i]);
    }
  }

  cout << dp[1];

  return 0;
}